File: lintcode/lintcode_376_Tree_pathSum.py
"""
Definition of TreeNode:
class TreeNode:
    def __init__(self, val):
        self.val = val
        self.left, self.right = None, None
"""


class Solution:
    # @param {TreeNode} root the root of binary tree
    # @param {int} target an integer
    # @return {int[][]} all valid paths
    def binaryTreePathSum(self, root, target):
        # Write your code here
        res = []
        temp = []
        if (root is None):
            return res
        res, temp = self.helper(root, target, res, temp)
        return res

    def helper(self, root, target, res, temp):
        if (root is None):
            return res, temp
        temp.append(root.val)
        if (root.val == target
                and root.left is None
                and root.right is None):
            res.append(temp[:])
        target -= root.val
        res, temp = self.helper(root.left, target, res, temp)
        res, temp = self.helper(root.right, target, res, temp)
        return res, temp[0 : -1]

File: lintcode/lintcode_376_Tree_pathSum.java
// Author: Xingxing Huang
// 注意：  传地址而不是传值
//        递归后面需要删除helper元素的存储的最后一位  
/**
 * Definition of TreeNode:
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left, right;
 *     public TreeNode(int val) {
 *         this.val = val;
 *         this.left = this.right = null;
 *     }
 * }
 */
public class Solution {
    /**
     * @param root the root of binary tree
     * @param target an integer
     * @return all valid paths
     */
    public List<List<Integer>> binaryTreePathSum(TreeNode root, int target) {
        // Write your code here
        List<List<Integer>> result = new ArrayList<List<Integer>>();
        List<Integer> list = new ArrayList<Integer>();
        if (root == null) { 
            return result;
        }
        pathSum(root, target, result, list);
        return result;
    }
    public void pathSum(TreeNode root, 
                        int target, 
                        List<List<Integer>> result, 
                        List<Integer> list) {
        if (root == null) {
            return;
        }
        list.add(root.val);
        if (root.val == target && root.left == null && root.right == null) {
            result.add(new ArrayList<Integer>(list));
        }
        target -= root.val;
        pathSum(root.left, target, result, list);
        pathSum(root.right, target, result, list);
        list.remove(list.size() - 1);
    }
}